packages/dds/tree/src/codec/codec.ts

```diff
@@ -0,0 +1,177 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { assert, bufferToString, IsoBuffer } from "@fluidframework/common-utils";
+import { fail, JsonCompatibleReadOnly } from "../util";
+
+/**
+ * @alpha
+ */
+export interface IEncoder<TDecoded, TEncoded> {
+	/**
+	 * Encodes `obj` into some encoded format. Typically paired with an {@link IDecoder}.
+	 */
+	encode(obj: TDecoded): TEncoded;
+}
+
+/**
+ * @alpha
+ */
+export interface IDecoder<TDecoded, TEncoded> {
+	/**
+	 * Decodes `obj` from some encoded format. Typically paired with an {@link IEncoder}.
+	 */
+	decode(obj: TEncoded): TDecoded;
+}
+
+/**
+ * @alpha
+ */
+export interface IJsonCodec<TDecoded, TEncoded = JsonCompatibleReadOnly>
```

Comment:
Wouldn't it make more sense to have an extends constraint here or just remove the option entirely?

Reply:
If not, then this type prabably needs  implementations to follow some addational ruls which go in the doc comment